dev-tools/omdb/src/bin/omdb/mgs/sensors.rs

```diff
@@ -0,0 +1,954 @@
+
+    reader.seek(position.clone())?;
+    let mut iter = reader.records();
+
+    let mut time = None;
+
+    loop {
+        *position = iter.reader().position().clone();
+
+        if let Some(record) = iter.next() {
+            let record = record?;
+
+            if record.len() != hlen {
+                bail!("bad record length at line {}", position.line());
+            }
+
+            let now = record[0].parse::<u64>().or_else(|_| {
+                bail!("bad time at line {}", position.line());
+            })?;
+
+            if let Some(time) = time {
+                if now != time {
+                    break;
+                }
+            } else {
+                if let Some(end) = metadata.end_time {
+                    if now > end {
+                        time = Some(0);
+                        break;
+                    }
+                }
+
+                time = Some(now);
+            }
+
+            if let Some(sensor) = Sensor::from_string(&record[1], &record[2]) {
+                if let Some(ids) = metadata.sensors_by_sensor.get_vec(&sensor) {
+                    for id in ids {
+                        let (_, _, d) = metadata.sensors_by_id.get(id).unwrap();
+                        let value = match record[d.field()].parse::<f32>() {
+                            Ok(value) => Some(value),
+                            _ => None,
+                        };
+
+                        values.insert(*id, value);
+                    }
+                }
+            } else {
+                bail!("bad sensor at line {}", position.line());
+            }
+        } else {
+            time = Some(0);
+            break;
+        }
+    }
+
+    Ok(SensorValues { values, latencies: None, time: time.unwrap() })
+}
+
+pub(crate) async fn sensor_data<R: std::io::Read + std::io::Seek>(
+    input: &mut SensorInput<R>,
+    metadata: &'static SensorMetadata,
+) -> Result<SensorValues, anyhow::Error> {
+    match input {
+        SensorInput::MgsClient(ref mgs_client) => {
+            sp_data_mgs(mgs_client, metadata).await
+        }
+        SensorInput::CsvReader(reader, position) => {
+            sp_data_csv(reader, position, metadata)
+        }
+    }
+}
+
+///
+/// Runs `omdb mgs sensors`
+///
+pub(crate) async fn cmd_mgs_sensors(
+    omdb: &crate::Omdb,
+    log: &slog::Logger,
+    mgs_args: &crate::mgs::MgsArgs,
+    args: &SensorsArgs,
+) -> Result<(), anyhow::Error> {
+    let mut input = if let Some(ref input) = args.input {
+        let file = File::open(input)?;
+        SensorInput::CsvReader(
+            csv::Reader::from_reader(file),
+            csv::Position::new(),
+        )
+    } else {
+        SensorInput::MgsClient(mgs_args.mgs_client(omdb, log).await?)
+    };
+
+    let (metadata, mut values) = sensor_metadata(&mut input, args).await?;
+
+    //
+    // A bit of shenangians to force metadata to be 'static -- which allows
+    // us to share it with tasks.
```

Comment:
We don't need to leak this to share it with tasks - the general thing here is to stash it into an `Arc`. The changes are spread out but pretty trivial:

```diff
--- a/dev-tools/omdb/src/bin/omdb/mgs/dashboard.rs
+++ b/dev-tools/omdb/src/bin/omdb/mgs/dashboard.rs
@@ -486,7 +486,7 @@ struct Dashboard {
 }

 impl Dashboard {
-    fn new(metadata: &'static SensorMetadata) -> Result<Dashboard> {
+    fn new(metadata: &SensorMetadata) -> Result<Dashboard> {
         let mut sps =
             metadata.sensors_by_sp.keys().map(|m| *m).collect::<Vec<_>>();
         let mut graphs = HashMap::new();
@@ -806,14 +806,7 @@ pub(crate) async fn cmd_mgs_dashboard(
     let (metadata, values) =
         sensor_metadata(&mut input, &args.sensors_args).await?;

-    //
-    // A bit of shenangians to force metadata to be 'static -- which allows
-    // us to share it with tasks.
-    //
-    let metadata = Box::leak(Box::new(metadata));
-    let metadata: &_ = metadata;
-
-    let mut dashboard = Dashboard::new(metadata)?;
+    let mut dashboard = Dashboard::new(&metadata)?;
     let mut last = values.time;
     let mut force = true;
     let mut update = true;
@@ -822,7 +815,7 @@ pub(crate) async fn cmd_mgs_dashboard(

     if args.sensors_args.input.is_some() && !args.simulate_realtime {
         loop {
-            let values = sensor_data(&mut input, metadata).await?;
+            let values = sensor_data(&mut input, &metadata).await?;

             if values.time == 0 {
                 break;
@@ -862,7 +855,7 @@ pub(crate) async fn cmd_mgs_dashboard(

         if update && now != last {
             let kicked = Instant::now();
-            let f = sensor_data(&mut input, metadata);
+            let f = sensor_data(&mut input, &metadata);
             last = now;

             while Instant::now().duration_since(kicked).as_millis() < 800 {
--- a/dev-tools/omdb/src/bin/omdb/mgs/sensors.rs
+++ b/dev-tools/omdb/src/bin/omdb/mgs/sensors.rs
@@ -15,6 +15,7 @@ use gateway_client::types::SpType;
 use multimap::MultiMap;
 use std::collections::{HashMap, HashSet};
 use std::fs::File;
+use std::sync::Arc;
 use std::time::{Duration, SystemTime, UNIX_EPOCH};

 #[derive(Debug, Args)]
@@ -543,7 +544,7 @@ fn sp_info_csv<R: std::io::Read>(
 pub(crate) async fn sensor_metadata<R: std::io::Read>(
     input: &mut SensorInput<R>,
     args: &SensorsArgs,
-) -> Result<(SensorMetadata, SensorValues), anyhow::Error> {
+) -> Result<(Arc<SensorMetadata>, SensorValues), anyhow::Error> {
     let by_kind = if let Some(types) = &args.types {
         let mut h = HashSet::new();

@@ -634,7 +635,7 @@ pub(crate) async fn sensor_metadata<R: std::io::Read>(
     }

     Ok((
-        SensorMetadata {
+        Arc::new(SensorMetadata {
             sensors_by_sensor,
             sensors_by_sensor_and_sp,
             sensors_by_id,
@@ -648,7 +649,7 @@ pub(crate) async fn sensor_metadata<R: std::io::Read>(
                     None => None,
                 },
             },
-        },
+        }),
         SensorValues { values, time, latencies: info.latencies },
     ))
 }
@@ -692,7 +693,7 @@ async fn sp_read_sensors(

 async fn sp_data_mgs(
     mgs_client: &gateway_client::Client,
-    metadata: &'static SensorMetadata,
+    metadata: &Arc<SensorMetadata>,
 ) -> Result<SensorValues, anyhow::Error> {
     let mut values = HashMap::new();
     let mut latencies = HashMap::new();
@@ -701,9 +702,10 @@ async fn sp_data_mgs(
     for sp_id in metadata.sensors_by_sp.keys() {
         let mgs_client = mgs_client.clone();
         let id = *sp_id;
+        let metadata = Arc::clone(&metadata);

         let handle = tokio::spawn(async move {
-            sp_read_sensors(&mgs_client, &id, metadata).await
+            sp_read_sensors(&mgs_client, &id, &metadata).await
         });

         handles.push((id, handle));
@@ -729,7 +731,7 @@ async fn sp_data_mgs(
 fn sp_data_csv<R: std::io::Read + std::io::Seek>(
     reader: &mut csv::Reader<R>,
     position: &mut csv::Position,
-    metadata: &'static SensorMetadata,
+    metadata: &SensorMetadata,
 ) -> Result<SensorValues, anyhow::Error> {
     let headers = reader.headers()?;
     let hlen = headers.len();
@@ -795,14 +797,14 @@ fn sp_data_csv<R: std::io::Read + std::io::Seek>(

 pub(crate) async fn sensor_data<R: std::io::Read + std::io::Seek>(
     input: &mut SensorInput<R>,
-    metadata: &'static SensorMetadata,
+    metadata: &Arc<SensorMetadata>,
 ) -> Result<SensorValues, anyhow::Error> {
     match input {
         SensorInput::MgsClient(ref mgs_client) => {
             sp_data_mgs(mgs_client, metadata).await
         }
         SensorInput::CsvReader(reader, position) => {
-            sp_data_csv(reader, position, metadata)
+            sp_data_csv(reader, position, &metadata)
         }
     }
 }
@@ -828,13 +830,6 @@ pub(crate) async fn cmd_mgs_sensors(

     let (metadata, mut values) = sensor_metadata(&mut input, args).await?;

-    //
-    // A bit of shenangians to force metadata to be 'static -- which allows
-    // us to share it with tasks.
-    //
-    let metadata = Box::leak(Box::new(metadata));
-    let metadata: &_ = metadata;
-
     let mut sensors = metadata.sensors_by_sensor.keys().collect::<Vec<_>>();
     sensors.sort();

@@ -939,7 +934,7 @@ pub(crate) async fn cmd_mgs_sensors(
             wakeup += tokio::time::Duration::from_millis(1000);
         }

-        values = sensor_data(&mut input, metadata).await?;
+        values = sensor_data(&mut input, &metadata).await?;

         if args.input.is_some() && values.time == 0 {
             break;
```

Reply:
Thank you for this! I had tried to get this working earlier, but had (clearly) failed!